### PR TITLE
Create user model

### PR DESCRIPTION
**Pull request by Ichenn92**
To add the user and his validations on the time without timestamping I added a gem name "tod" which stands for Time Of Day.
Contains tones of helper for periods of time. the overlap between 2 schedules etc...

I added some tests for confirmation on the model and in case we need to change it.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -3,4 +3,22 @@ class User < ApplicationRecord
...
+
+  validates_presence_of :nickname, :first_name, :last_name
+  validates_inclusion_of :actor, in: [true, false]
+
+  validate :start_time_cannot_be_greater_or_equal_end_time, if: :actor?
+
+  with_options if: :actor? do
+    validates :availibility_start_time, presence: true
+    validates :availibility_end_time, presence: true
+  end
```
Comment (michaellaurac) [inline app/models/user.rb:17]: Move line 12 in the block with the if condition on actor

File: app/models/user.rb [hunk 2]
```diff
@@ -3,4 +3,22 @@ class User < ApplicationRecord
...
+
+  def start_time_cannot_be_greater_or_equal_end_time
+    if availibility_start_time > availibility_end_time || availibility_start_time == availibility_end_time
+      errors.add(:availibility_start_time, "can't be later or equal to end time")
```
Comment (michaellaurac) [inline app/models/user.rb]: modify
errors.add(:availability_start_time, "must be earlier to availability end time")
errors.add(:availability_end_time, "must be later to availability end time")

File: test/models/user_test.rb [hunk 3]
```diff
@@ -1,7 +1,42 @@
...
+  end
+
+  test "validates start availibility time is earlier than end availibility time" do
+    u = @@base_valid_actor
+    u.availibility_start_time = Tod::TimeOfDay.new 8, 15
+    u.availibility_end_time = Tod::TimeOfDay.new 16, 45
+    assert u.valid?, u.errors.full_messages.inspect
+  end
+
+  test "reject when start availibility time is later to end availibility time" do
```
Comment (michaellaurac) [inline test/models/user_test.rb:29]: Later then

File: test/models/user_test.rb [hunk 4]
```diff
@@ -1,7 +1,42 @@
...
-  # end
+  @@base_valid_actor = User.new(nickname: "titi",
+                                first_name: "toto",
+                                last_name: "tete",
+                                email: "titi@tete.ch",
+                                actor: true,
+                                password: "12341234")
+
+  test "User.actor? default is false" do
+    u = User.new
```
Comment (michaellaurac) [inline test/models/user_test.rb:12]: rename 'u' as 'test_user' if possible

**Review by clem13-coder — COMMENTED** [review]
See Michael Comment

**Review by clem13-coder — APPROVED** [review]

**Review by clem13-coder — APPROVED** [review]

**Review by clem13-coder — APPROVED** [review]

File: app/models/user.rb [hunk 5]
```diff
@@ -3,4 +3,23 @@ class User < ApplicationRecord
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
+  serialize :availibility_start_time, Tod::TimeOfDay
+  serialize :availibility_end_time, Tod::TimeOfDay
+
+  validates_presence_of :nickname, :first_name, :last_name
+  validates_inclusion_of :actor, in: [true, false]
+
+  validate :start_time_cannot_be_greater_or_equal_end_time, if: :actor?
```
Comment (michaellaurac) [inline app/models/user.rb:12]: Check later if validates_with MY_VALIDATOR

**Review by michaellaurac — APPROVED** [review]

**Pull request merged**